src/lib.rs

```diff
@@ -301,4 +368,54 @@ impl MemoryMappedDevice {
             chunk.copy_from_slice(&bytes[..chunk.len()]);
         }
     }
+
+    fn start_dma(&self, access: &FWCfgDmaAccess) {
+        let address = access as *const FWCfgDmaAccess as u64;
+        // https://gitlab.com/qemu-project/qemu/-/blob/v7.0.0/docs/specs/fw_cfg.txt#L89
+        let offset = 16;
+        let dma_address_register: *mut u32 = self.register(offset);
+        unsafe {
+            // https://gitlab.com/qemu-project/qemu/-/blob/v7.0.0/docs/specs/fw_cfg.txt#L167
+            // The DMA address register is 64-bit and big-endian.
+            // Writing its lower half is what triggers DMA,
+            // so write these half separately to control their order:
+            let register_high = dma_address_register;
+            let register_low = dma_address_register.add(1); // One u32
+            let address_high = (address >> 32) as u32;
+            let address_low = address as u32;
+            register_high.write_volatile(address_high.to_be());
+            compiler_fence(Ordering::AcqRel);
+            register_low.write_volatile(address_low.to_be());
+        }
+    }
+}
+
+#[derive(Debug)]
+// NOTE: The memory layout of this struct must match this exactly:
+// https://gitlab.com/qemu-project/qemu/-/blob/v7.0.0/docs/specs/fw_cfg.txt#L177-181
+#[repr(C)]
+struct FWCfgDmaAccess {
```

Comment:
Should be `FwCfgDmaAccess`, to keep things consistent.

Reply:
Done!